numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1810,39 +1835,75 @@ unpack_bits(PyObject *input, int axis, PyObject *count_obj, char order)
         unsigned const char *inptr = PyArray_ITER_DATA(it);
         char *outptr = PyArray_ITER_DATA(ot);
 
-        if (order == 'b') {
-            for (index = 0; index < in_n; index++) {
-                for (i = 0; i < 8; i++) {
-                    *outptr = ((*inptr & (128 >> i)) != 0);
-                    outptr += out_stride;
+        if (out_stride == 1) {
+            /* for unity stride we can just copy out of the lookup table */
+            if (order == 'b') {
+                for (index = 0; index < in_n; index++) {
+                    npy_uint64 v = unpack_lookup_big[*inptr].uint64;
+                    memcpy(outptr, &v, 8);
+                    outptr += 8;
+                    inptr += in_stride;
+                }
+            }
+            else {
+                for (index = 0; index < in_n; index++) {
+                    npy_uint64 v = unpack_lookup_big[*inptr].uint64;
+                    if (order != 'b') {
+                        v = npy_bswap8(v);
+                    }
+                    memcpy(outptr, &v, 8);
+                    outptr += 8;
+                    inptr += in_stride;
                 }
-                inptr += in_stride;
             }
             /* Clean up the tail portion */
-            for (i = 0; i < in_tail; i++) {
-                *outptr = ((*inptr & (128 >> i)) != 0);
-                outptr += out_stride;
+            if (in_tail) {
+                npy_uint64 v = unpack_lookup_big[*inptr].uint64;
+                if (order != 'b') {
+                    v = npy_bswap8(v);
+                }
+                memcpy(outptr, &v, in_tail);
+            }
+            /* Add padding */
+            else if (out_pad) {
+                memset(outptr, 0, out_pad);
             }
         }
         else {
-            for (index = 0; index < in_n; index++) {
-                for (i = 0; i < 8; i++) {
+            if (order == 'b') {
+                for (index = 0; index < in_n; index++) {
+                    for (i = 0; i < 8; i++) {
```

Comment:
Wouldn't it make sense to still use the lookup table even for stride != 1 (i.e., replace the `memcpy`, not the table reading).

Reply:
probably, I never bothered testing as it is probably an edge case, though as we are updating it we might as well go all the way

Reply:
If the proposed changes will take more than a day or two, perhaps we should merge this as-is with a comment for future improvements, since the deadline for 1.17 is soon

Reply:
I probably won't get to do it, imo this can be merged now

Reply:
Anything holding this up? Otherwise I will make a pass and merge soon (or probably Matti beats me to it ;)).

Reply:
LGTM. The suggestion to use the table for the strided case is good, but can go in later, we need the fix for 1.17.